src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataReaderExtensions.cs

```diff
@@ -71,7 +60,26 @@ public static ODataItemBase ReadResourceOrResourceSet(this ODataReader reader)
                             "The resource which is ending should be on the top of the items stack.");
                         itemsStack.Pop();
                         break;
+                    case ODataReaderState.DeletedResourceStart:
+                        ODataDeletedResource deletedResource = (ODataDeletedResource)reader.Item;
+                        ODataResourceWrapper deletedResourceWrapper = null;
+                        Contract.Assert(deletedResource != null, "Deleted resource should not be null");
 
+                        deletedResourceWrapper = new ODataResourceWrapper(deletedResource);
+
+                        Contract.Assert(itemsStack.Count != 0, "Deleted Resource should not be top level item");
+
+                        AddResourceToParent(itemsStack, deletedResourceWrapper);
+
+                        itemsStack.Push(deletedResourceWrapper);
+                        break;
+
+                    case ODataReaderState.DeletedResourceEnd:
+                        Contract.Assert(
+                            itemsStack.Count > 0 && (itemsStack.Peek().Item == reader.Item),
+                            "The resource which is ending should be on the top of the items stack.");
+                        itemsStack.Pop();
+                        break;
```

Comment:
just have this case fall through to the existing ResourceEnd state on line 57:

                    case ODataReaderState.ResourceEnd:
                    case ODataReaderState.DeletedResourceEnd:
                        Contract.Assert(
                            itemsStack.Count > 0 && (reader.Item == null || itemsStack.Peek().Item == reader.Item),
                            "The resource which is ending should be on the top of the items stack.");
                        itemsStack.Pop();
                        break;